Improvements to slotting conversions / TimeInterpreter

Anviking opened the pull request:
# Issue Number

#1869 / #1960 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] Block instead of using a `singleEraInterpreter` if we don't yet have a proper `TimeInterpreter`.
- [x] Remove TODO; explain why delegation time conversion never should fail
- [x] Replace `TimeInterpreter IO` with `TimeInterpreter (ExceptT IO ErrPastHorizon)`
- [x] Add, test and use `endTimeOfEpoch e` instead of `timeOf =<< firstSlotInEpoch (e+1)`. This is needed to stay inside the forecast range when showing `next_epoch` when the node is in-sync, in Byron. Fixes #1960 
- [x] Make `next_epoch` and `network_tip` optional in the API. When the node is still syncing the Byron chain, it cannot know any slotting info close to the current time. Solving related problem to #1960 
- [x] Revise the names of the constructors of `TimeInterpreter`


# Comments

- I thought queries in Shelley could fail. This is wrong. Once in Shelley (for now the final era), they have unlimited forecast range. However, if the node is in Byron, we need to be careful.
<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Use MVar instead of TVar (Maybe ) to store TimeInterpreter

On launch, the TimeInterpreter may not have been fetched from the node.

Instead returning the singleEraInterpreter for the first era, it seems
safer to block until fetched.

I imagine there could be race conditions where we would sometimes return
completely wrong time data when in Shelley, just after starting the
node. Not completely sure, but I hope there shouldn't be any drawbacks
with blocking.

Anviking committed:
Explain why delegation time conversion never should fail

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -456,10 +456,10 @@ withNetworkLayer tr np addrInfo versionData action = do
-        cached <- atomically (readTVar var)
-        let interpret = maybe (interpreterFromGenesis gp) (mkTimeInterpreter getGenesisBlockDate) cached
+        cached <- takeMVar var
+        let interpret = (mkTimeInterpreter getGenesisBlockDate) cached
```
I am quite failing to grasp the consequences of this.

Anviking replied:
Goal:
```
On launch, the TimeInterpreter may not have been fetched from the node.

Instead returning the singleEraInterpreter for the first era, it seems
safer to block until fetched.

I imagine there could otherwise be race conditions where we would sometimes
return completely wrong time data when in Shelley, just after starting the
node.
```
But:
- I should use `readMVar` instead of `takeMVar` 😅

Anviking committed:
Simplify constructors of TimeInterpreter

Anviking committed:
Use TimeInterpreter (ExceptT PastTimeHorizon IO)

Anviking committed:
Add endTimeOfEpoch and test

Anviking committed:
Make network_tip and next_epoch optional

If the node is not enough in-sync, we cannot know them.

Anviking committed:
Use TimeInterpreter for pool retirements

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -450,15 +461,18 @@ withNetworkLayer tr np addrInfo versionData action = do
             Left e -> do
-                traceWith tr $ MsgInterpreterPastHorizon (pretty query) e
-                throwIO e
+                liftIO $ traceWith tr $ MsgInterpreterPastHorizon (pretty query) e
```
`MsgInterpreterPastHorizon` shouldn't be critical anymore, as it might be expected.

Anviking replied:
I lowered to `Info` and removed the log line about it being unexpected.

Anviking committed:
Lower MsgInterpreterPastHorizon to Error

Anviking committed:
*If* the time of the tip ever fails, don't crash

But rather respond with NotResponding.

Anviking committed:
Fix overflow in syncProgress calculation

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1349,8 +1366,9 @@ joinStakePool ctx knownPools getPoolStatus apiPoolId (ApiT wid) body = do
+    -- Not forecasting into the future. Should be safe.
     ti :: TimeInterpreter IO
-    ti = timeInterpreter (ctx ^. networkLayer @t)
+    ti = unsafeRunExceptT . timeInterpreter (ctx ^. networkLayer @t)
```
It doesn't seem optimal to have to argue "should be safe". But I'm not sure that is easy to solve for the moment.

KtorZ replied:
`unsafe` will throw and cause a 500, we shouldn't use it if that can indeed happen. What about instead having a proper error, even if it's just a server error, but at least not an opaque 500.

Anviking replied:
> instead having a proper error

I realised that we should at least use a custom error message instead of `unsafeRunExceptT`, and have the motivation why it is impossible _there_ instead of in a comment.

Anviking committed:
Add regression test for syncProgress overflow

It fails without the recent fix

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -26,13 +26,14 @@ module Cardano.Wallet.Primitive.Slotting
-    , interpreterFromGenesis
     , mkTimeInterpreter
-    , MyInterpreter(..)
+    , unsafeNoForecasts
```
I should either use this consistently over `unsafeRunExceptT`, or remove it.

Anviking committed:
remove 'unsafeRunExceptT' calls on the time interpreter

  Instead, we do not push the exception down to every caller but rather, throw it as an exception in the network layer. The rationale is that, this exception can only occur when both conditions are met:

  a) The node is still syncing and doesn't yet know about any hard-fork.
  b) A time beyond the node's foreseeable future is queried.

  While syncing in Byron these two conditions can't be met (times referenced in blocks are neccessarily before the node's tip and can't be beyond its foreseeable future. There's the case of delegation certificates and or transaction TTL but these only exists in Shelley, where the foreseeable future is so far, unlimited.
  Yet, there are points in the API where a time that is far beyond the node's tip can be provided and that is:

  - As filtering parameter when listing transactions.
  - As current time when looking at network parameters.

rvl reviewed (commented):
There is so much `unsafeRunExceptT` here. It makes our code look unsafe.
This is replacing an actual exception type `PastHorizonException` with `userError` and a string message, which is worse.

Why not let all the time/slot conversion functions throw PastHorizonException, wherever unsafeRunExceptT is currently used. In the other "safe" cases, catch the exception.

Alternatively, replace `unsafeRunExceptT` with a new util `runNonFailingQuery` (or some name like that). This will be basically `runNonFailingQuery q = q >>= either throwIO pure`.  The advantage of this is that it's obvious where the exception could occur, and userError is not used.


And also we should sprinkle HasCallStack liberally - otherwise tracking down errors will be hard.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -1024,4 +1037,4 @@ instance HasSeverityAnnotation (NetworkLayerLog b) where
         MsgChainSyncCmd cmd                -> getSeverityAnnotation cmd
         MsgInterpreter{}                   -> Debug
-        MsgInterpreterPastHorizon{}        -> Critical
+        MsgInterpreterPastHorizon{}        -> Info
```
Is this change really sound :thinking: ?

rvl replied:
I think the reasoning is that we have actually some cases where the PastHorizonException is "normal" and somewhat expected.

But perhaps we can divide slot/time conversions into two categories:
1. Where it's our bug if the conversion fails.
2. Where any conversion errors are because of the user inputting a date which is beyond the safe zone. So an error not a bug.

KtorZ committed:
handle plausible PastTimeHorizon exception in the API server.
